src/actions.ts

```diff
@@ -0,0 +1,449 @@
+		}
+		if (!params.cpassword) {
+			throw new ActionError(`Repeat your new password.`);
+		}
+
+		if (!this.user.loggedin) {
+			throw new ActionError('Your session has expired. Please log in again.');
+		}
+		if (params.password !== params.cpassword) {
+			throw new ActionError('Your new passwords do not match.');
+		}
+		if (!(await this.session.passwordVerify(this.user.id, params.oldpassword))) {
+			throw new ActionError('Your old password was incorrect.');
+		}
+		params.password = params.password.replace(/\s/ig, '');
+		if (params.password.length < 5) {
+			throw new ActionError('Your new password must be at least 5 characters long.');
+		}
+		const actionsuccess = await this.session.changePassword(this.user.id, params.password);
+		return {actionsuccess};
+	},
+	async changeusername(params) {
+		if (this.request.method !== 'POST') {
+			throw new ActionError('Invalid request (username changing must be done through POST requests).');
+		}
+		if (!params.username) {
+			throw new ActionError(`Specify a username.`);
+		}
+		if (!this.user.loggedin) {
+			throw new ActionError('Your session has expired. Please log in again.');
+		}
+		if (toID(params.username) !== this.user.id) {
+			throw new ActionError('You\'re not logged in as that user.');
+		}
+		// safe to use userid directly because we've confirmed they've logged in.
+		const actionsuccess = await tables.users.update(this.user.id, {
+			username: params.username,
+		}).catch(() => false);
+		await this.session.setSid();
+		return {actionsuccess};
+	},
+	async getassertion(params) {
+		this.setPrefix('');
+		params.userid = toID(params.userid);
+		if (!params.userid) {
+			params.userid = this.user.id;
+		}
+		if (params.userid === 'guest') {
+			return ';'; // Special error message for this case.
+		}
+		// NaN is falsy so this validates
+		const challengekeyid = Number(params.challengekeyid) || -1;
+		const challenge = params.challenge || params.challstr || "";
+		return this.session.getAssertion(
+			params.userid,
+			challengekeyid,
+			this.user,
+			challenge,
+			this.verifyCrossDomainRequest()
+		);
+	},
+	async ladderupdate(params) {
+		const server = await this.getServer(true);
+		if (server?.id !== Config.mainserver) {
+			return {errorip: "Your version of PS is too old for this ladder system. Please update."};
+		}
+
+		if (!toID(params.format)) throw new ActionError("Invalid format.");
+		const ladder = new NTBBLadder(params.format);
+		const p1 = NTBBLadder.getUserData(params.p1);
+		const p2 = NTBBLadder.getUserData(params.p2);
+		if (!p1 || !p2) {
+			// The server should not send usernames > 18 characters long.
+			// (getUserData returns falsy when the usernames are too long)
+			return 0;
+		}
+
+		const out: {[k: string]: any} = {};
+		await ladder.updateRating(p1, p2, parseFloat(params.score));
+		out.actionsuccess = true;
+		out.p1rating = p1.rating;
+		out.p2rating = p2.rating;
+		delete out.p1rating.rpdata;
+		delete out.p2rating.rpdata;
```

Comment:
I don't know what the type of the return value is supposed to be. Don't use `any`. And instead of assigning everything after the fact and then deleting some data, maybe put the values you want to send directly in the object constructor

Reply:
I can't say I entirely understand what this is supposed to be either. This file I converted pretty directly from PHP since the rating code is nebulous to me. 

My best interpretation is that it's just the entire rating row, direct from the database? RPData definitely needs to be removed since it can be massive - no point wasting bandwidth on it.